src/unitxt/metrics.py

```diff
@@ -400,31 +699,180 @@ def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generato
+                "score": {
+                    "instance": {
+                        field_name: group_aggregation_func(scores[field_name])
+                        for field_name in field_names
+                    }
+                }
+            }
+            for scores in group_to_instance_scores.values()
+        ]
+
+    def get_group_aggregated_instance_scores_by_subgroups(
+        self,
+        instances: List[dict],
+        field_names: List[str],
+        group_aggregation_func,
+        subgroup_field: str,
+        baseline_name,
+    ):
+        """Return a list of group aggregation function value for group_mean_subgroup_comparison reduction.
+
+        Args:
+            instances: List of observation instances with instance-level scores (fields) computed.
+            field_names: List of instance score names in each instance to apply the aggregation function.
+            group_aggregation_func: Callable aggregation function with arguments baseline_scores and other_scores.
+            subgroup_field: name of field in instance additional_inputs that contains the subgroup identifier
+            baseline_name: value of subgroup_field that indicates a score belongs in baseline_scores.
+
+        Returns:
+            List of dicts, each corresponding to a group of instances (defined by grouping_field)
+        """
+        from collections import defaultdict
+
+        # two-level defaultdict: first is the grouping, second is the field name
+        # first list is instance scores for baseline group, second is for comparison group
+        group_to_instance_scores = defaultdict(lambda: defaultdict(lambda: [[], []]))
 
         for instance in instances:
-            yield instance
+            additional_inputs = instance["additional_inputs"]
+            for cc in [self.grouping_field, subgroup_field]:
+                if cc not in additional_inputs:
+                    raise ValueError(
+                        f"Missing '{cc}' from instance {instance}. "
+                        f"This field is required for group based metric computation."
+                    )
+            group_key = additional_inputs[self.grouping_field]
+            # indicator if is in the baseline group or not
+            is_baseline = str(additional_inputs[subgroup_field]) == baseline_name
+            # convert True (baseline) to 0, and False (others) to 1, store in respective groups
+            idx = int(not is_baseline)
+
+            for field_name in field_names:
+                group_to_instance_scores[group_key][field_name][idx].append(
+                    instance["score"]["instance"][field_name]
+                )
+
+        # now for each group, take the aggregation function, comparing others to baseline
+        return [
+            {
+                "score": {
+                    "instance": {
+                        field_name: group_aggregation_func(
+                            baseline_scores=scores[field_name][0],
```

Comment:
was wondering if perhaps the separation between grouping with subgroups and without subgroups could be avoided: 
1. assume we always have subgroups (baselines and other scores). 
2. The dataset will define what are the baselines and what are the "others" (as coded now). In some cases -- there will be no baselines defined in the dataset. 
3. The `group_aggregation_func`s API will always accept `baseline_scores` and `other_scores`, but it should verify the input is as expected. For example, if we just do mean aggregation for a group, the function should make sure there are no baselines (otherwise assert). If we do PDR, make sure there is a baseline given (otherwise assert), and so on... 

The main advantage of that would be IMO that most of the code can be shared between the two group aggregation options, and it would be more easy to follow.. At the cost of some added verification code for the group aggregation functions.

Reply:
Possibly, but the subgroup aggregation functions require additional arguments.  I see that one could check if there is a "is_baseline" column and if not, define it to always be True.  Maybe in these cases it is a good idea to create a copy of the instances so we don't add extra fields to the list that is returned.  But then we have to check the arguments of aggregation functions etc.